src/instantiate.jl

```diff
@@ -0,0 +1,104 @@
+"""
+    struct UninstantiatedOptimizer
+        optimizer
+        params::Vector{Pair{AbstractOptimizerAttribute,<:Any}}
+    end
+
+Object representing an optimizer, created by `optimizer()` parametrized by
+parameters listed in `params`. Instances are created with [`instantiate`](@ref).
+"""
+struct UninstantiatedOptimizer
+    # Function that takes zero arguments and returns a new optimizer.
+    # The type of the function could be
+    # * `Function`: a function, or
+    # * `DataType`: a type, or
+    # * `UnionAll`: a type with missing parameters.
+    optimizer
+    params::Vector{Pair{AbstractOptimizerAttribute,Any}}
+end
+
+_to_param(param::Pair{<:AbstractOptimizerAttribute}) = param
+_to_param(param::Pair) = RawParameter(param.first) => param.second
+_to_param(param::Pair{Symbol}) = _to_param(string(param.first) => param.second)
+function _add_params(params::Vector{Pair{AbstractOptimizerAttribute,Any}}) end
+function _add_params(params::Vector{Pair{AbstractOptimizerAttribute,Any}},
+                     param::Pair, other_params::Vararg{Pair, N}) where N
+    push!(params, _to_param(param))
+    _add_params(params, other_params...)
+end
+
+"""
+    parametrize(uninstantiated, params::Pair...)
```

Comment:
To parametrize means to describe in terms of parameters, i.e., listing all possible parameters. This is different in my mind from specifying some subset of parameters.
I would suggest `optimizer_with_parameters`.